lib/browser/http.ts

```diff
@@ -193,3 +194,109 @@ export class HttpClientStream extends BufferedEventEmitter {
+    private pending_requests: PendingRequest[] = [];
+
+    constructor(
+        readonly host: string,
+        readonly port: number,
+        readonly max_connections: number
+    ) {
+        
+    }
+
+    private remove(connection: HttpClientConnection) {
+        this.pending_connections.delete(connection);
+        this.live_connections.delete(connection);
+        const free_idx = this.free_connections.indexOf(connection);
+        if (free_idx != -1) {
+            this.free_connections.splice(free_idx, 1);
+        }
+    }
+
+    private resolve(connection: HttpClientConnection) {
+        const request = this.pending_requests.shift();
+        if (request) {
+            request.resolve(connection);
+        }
+    }
+
+    private reject(error: CrtError) {
+        const request = this.pending_requests.shift();
+        if (request) {
+            request.reject(error);
+        }
+    }
+
+    private pump() {
+        // Try to service the request with a free connection
+        {
+            let connection = this.free_connections.pop();
+            if (connection) {
+                return this.resolve(connection);
+            }
+        }
+        
+        // If there's no more room, nothing can be resolved right now
+        if ((this.live_connections.size + this.pending_connections.size) == this.max_connections) {
+            return;
+        }
+        
+        // There's room, create a new connection
+        let connection = new HttpClientConnection(this.host, this.port);
+        this.pending_connections.add(connection);
+        const on_connect = () => {
+            this.pending_connections.delete(connection);
+            this.live_connections.add(connection);
+            this.free_connections.push(connection);
+            this.resolve(connection);
+        }
+        const on_error = (error: any) => {
+            if (this.pending_connections.has(connection)) {
+                // Connection never connected, error it out
+                return this.reject(new CrtError(error));
+            }
+            // If the connection errors after use, get it out of rotation and replace it
+            this.remove(connection);
+            this.pump();
+        }
+        const on_close = () => {
+            this.remove(connection);
+            this.pump();
+        }
+        connection.on('connect', on_connect);
+        connection.on('error', on_error);
+        connection.on('close', on_close);
+    }
+
+    acquire(): Promise<HttpClientConnection> {
+        return new Promise((resolve, reject) => {
+            this.pending_requests.push({
+                resolve: resolve,
+                reject: reject
+            });
+            this.pump();
+        });
+    }
+
+    release(connection: HttpClientConnection) {
+        this.free_connections.push(connection);
+        this.pump();
+    }
+
+    close() {
+        this.pending_requests.forEach((request) => {
```

Comment:
Should this attempt to do something with the connections and pending connections as well?

Reply:
Nope, the connections just go out of scope and close themselves. The test proves that by not hanging, since any open handles cause node to hang.